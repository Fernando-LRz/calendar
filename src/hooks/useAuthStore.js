import { useDispatch, useSelector } from 'react-redux';

import { clearErrorMessage, onChecking, onLogin, onLogout } from '../store';
import { calendarApi } from '../api';

export const useAuthStore = () => {

    const { status, user, errorMessage } = useSelector(state => state.auth);
    const dispatch = useDispatch();

    const startLogin = async({ email, password }) => {
        dispatch( onChecking() );

        try {
            const { data } = await calendarApi.post('/auth', { email, password });

            localStorage.setItem('token', data.token);
            localStorage.setItem('token-init-date', new Date().getTime());

            dispatch( onLogin({ name: data.name, uid: data.uid }) );
            
        } catch (error) {
            // console.log(error);
            const { response: { data } } = error;
            dispatch( onLogout(data.msg) );

            setTimeout(() => {
                dispatch( clearErrorMessage() );
            }, 10);
        }
    }

    return {
        errorMessage,
        status,
        user,

        startLogin,
    }
}