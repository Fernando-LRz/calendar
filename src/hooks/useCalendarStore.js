import { useDispatch, useSelector } from 'react-redux';

import { onAddNewEvent, onDeleteEvent, onLoadEvents, onSetActiveEvent, onUpdateEvent } from '../store';
import { convertEventsToDateEvents } from '../helpers';
import { calendarApi } from '../api';
import Swal from 'sweetalert2';

export const useCalendarStore = () => {

    const dispatch = useDispatch();
    const { events, activeEvent } = useSelector( state => state.calendar );
    const { user } = useSelector( state => state.auth );

    const setActiveEvent = (calendarEvent) => {
        dispatch( onSetActiveEvent(calendarEvent) );
    }

    const startSavingEvent = async(calendarEvent) => {
        try {
            // Update
            if(calendarEvent.id) {
                await calendarApi.put(`/events/${ calendarEvent.id }`, calendarEvent);
                dispatch( onUpdateEvent({ ...calendarEvent, user }) );
    
                return;
            } 
    
            // Create
            const { data } = await calendarApi.post('/events', calendarEvent);
            dispatch( onAddNewEvent({ ...calendarEvent, id: data.event.id, user }) );
            
        } catch (error) {
            // console.log(error);

            const { response: { data } } = error;
            Swal.fire('Error', data.msg, 'error');
        }
    }

    const startDeletingEvent = () => {
        dispatch( onDeleteEvent() );
    }

    const startLoadingEvents = async () => {
        try {
            const { data } = await calendarApi.get('/events');
            const events = convertEventsToDateEvents(data.events);

            dispatch( onLoadEvents(events) );

        } catch (error) {
            // console.log(error);
        }
    }

    return {
        events, 
        activeEvent,
        hasEventSelected: !!activeEvent,

        setActiveEvent,
        startSavingEvent,
        startDeletingEvent,
        startLoadingEvents
    }
}