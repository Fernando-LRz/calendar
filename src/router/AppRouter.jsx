import { useEffect } from 'react';
import { Navigate, Route, Routes } from 'react-router-dom';

import { LoginPage } from '../auth/';
import { CalendarPage } from '../calendar/';
import { useAuthStore } from '../hooks';

export const AppRouter = () => {

    const { checkAuthToken, status } = useAuthStore();

    useEffect(() => {
        checkAuthToken();
    }, []);

    if(status === 'checking') {
        return (
            <div className="d-flex justify-content-center align-items-center vh-100 bg-light">
                <h3 className="text-primary">Loading...</h3>
            </div>
        );
    }

    return (
        <Routes>
            
            {
                (status === "not-authenticated")
                    ? (
                        <>
                            <Route path="/auth/*" element={ <LoginPage /> } />
                            <Route path="/*" element={ <Navigate to="/auth/login" /> } />    
                        </>
                    )
                    : (
                        <>
                            <Route path="/" element={ <CalendarPage /> } />
                            <Route path="/*" element={ <Navigate to="/" /> } />    
                        </>
                    )
            }

        </Routes>
    )
}