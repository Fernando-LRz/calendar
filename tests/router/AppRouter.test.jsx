import { render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';

import { useAuthStore } from '../../src/hooks/useAuthStore';
import { AppRouter } from '../../src/router/AppRouter';

jest.mock('../../src/hooks/useAuthStore');

jest.mock('../../src/calendar/', () => ({
    CalendarPage: () => <h1>CalendarPage</h1>
}));

describe('<AppRouter /> tests', () => {

    const mockCheckAuthToken = jest.fn();
    beforeEach(() => jest.clearAllMocks());

    test('should show the loading screen and call the function to check the token', () => {

        useAuthStore.mockReturnValue({
            status: 'checking',
            checkAuthToken: mockCheckAuthToken
        });

        render( <AppRouter /> );
        
        expect( screen.getByText('Loading...') ).toBeTruthy();
        expect( mockCheckAuthToken ).toHaveBeenCalled();
    });

    test('should display the login if not authenticated', () => {

        useAuthStore.mockReturnValue({
            status: 'not-authenticated',
            checkAuthToken: mockCheckAuthToken
        });

        const { container } = render(
            <MemoryRouter>
                <AppRouter />
            </MemoryRouter>
        );

        // expect( screen.getByText('Log in') ).toBeTruthy();
        expect( container ).toMatchSnapshot();
    });

    test('should display the calendar if authenticated', () => {

        useAuthStore.mockReturnValue({
            status: 'authenticated',
            checkAuthToken: mockCheckAuthToken
        });

        render(
            <MemoryRouter>
                <AppRouter />
            </MemoryRouter>
        );

        expect( screen.getByText('CalendarPage') ).toBeTruthy();
    });

});