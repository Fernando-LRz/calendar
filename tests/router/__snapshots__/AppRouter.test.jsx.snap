// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AppRouter /> tests should display the login if not authenticated 1`] = `
<div>
  <div
    class="container login-container"
  >
    <div
      class="row"
    >
      <div
        class="col-md-6 login-form-1"
      >
        <h3>
          Log in
        </h3>
        <form>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="loginEmail"
              placeholder="Email"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="loginPassword"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <div
            class="d-grid gap-2"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Log in"
            />
          </div>
        </form>
      </div>
      <div
        class="col-md-6 login-form-2"
      >
        <h3>
          Sign up
        </h3>
        <form>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="signUpName"
              placeholder="Name"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="signUpEmail"
              placeholder="Email"
              type="email"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="signUpPassword"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="signUpConfirmedPassword"
              placeholder="Confirm your password"
              type="password"
              value=""
            />
          </div>
          <div
            class="d-grid gap-2"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Sign up"
            />
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
